### PR TITLE
[Analytics Engine] Document new string functions supported by the Analytics Engine SQL API

### PR DESCRIPTION
**Pull request by angelampcosta**

### CODE REVIEW

File: content/analytics/analytics-engine/sql-reference.md [hunk 1]
```diff
@@ -438,6 +438,86 @@ NOW()
...
+Extracts part of a string, starting at the Unicode code point indicated by the offset and returning the number of code points requested by the length. As previously mentioned, in SQL, indexes are usually 1-based. That means that the offset provided to substring should be at least 1.
+
+### format
+
+Usage:
+```SQL
+format({string}[, ...])
+```
+
+This function supports formatting strings, integers, floats, datetimes, intervals, etc, except NULL. We do not support literal `{ and }` in the format string and numbered arguments.
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md]: Made a quick correction, and I dropped the mention of numbered arguments as users won't necessarily expect them (easier to not mention.)

```suggestion
This function supports formatting strings, integers, floats, datetimes, intervals, etc, except NULL. We do not support literal `{` and `}` characters in the format string.
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 2]
```diff
@@ -438,6 +438,86 @@ NOW()
...
 
+### length
+
+Usage:
+```SQL
+length({string})
+```
+
+Returns the length of a string. This function is UTF-8 compatible.
+
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md:453]:
```suggestion
Examples:
```SQL
SELECT length('a string');
SELECT length(blob1) FROM your_dataset;
```
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 3]
```diff
@@ -438,6 +438,86 @@ NOW()
...
+### isEmpty
+
+Usage:
+```SQL
+isEmpty({string})
+```
+
+Returns a boolean saying whether the string was empty. This computation can also be done as a binary operation, `{string} = ''`.
+
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md:474]:
```suggestion
Examples:
```SQL
SELECT isEmpty('a string');
SELECT isEmpty(blob1) FROM your_dataset;
```
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 4]
```diff
@@ -438,6 +438,86 @@ NOW()
...
+### toLower
+
+Usage:
+```SQL
+toLower({string})
+```
+
+Returns the string converted to lowercase. This function is Unicode compatible. This may not be perfect for all languages and users with stringent needs, should do the operation in their own code.
+
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md:483]:
```suggestion
Examples:
```SQL
SELECT toLower('STRING TO DOWNCASE');
SELECT toLower(blob1) FROM your_dataset;
```
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 5]
```diff
@@ -438,6 +438,86 @@ NOW()
...
+### toUpper
+
+Usage:
+```SQL
+toUpper({string})
+```
+
+Returns the string converted to uppercase. This function is Unicode compatible. This may not be perfect for all languages and users with stringent needs, should do the operation in their own code.
+
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md:504]:
```suggestion
Examples:
```SQL
SELECT toUpper('string to uppercase');
SELECT toUpper(blob1) FROM your_dataset;
```
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 6]
```diff
@@ -438,6 +438,86 @@ NOW()
...
+### startsWith
+
+Usage:
+```SQL
+startsWith({string}, {string})
+```
+
+Returns a boolean of whether the first string has the second string at its start.
+
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md:513]:
```suggestion
Examples:
```SQL
SELECT startsWith('prefix ...', 'prefix');
SELECT startsWith(blob1, 'prefix') FROM your_dataset;
```
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 7]
```diff
@@ -438,6 +438,86 @@ NOW()
...
+### endsWith
+
+Usage:
+```SQL
+endsWith({string}, {string})
+```
+
+Returns a boolean of whether the first string contains the second string at its end.
+
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md:528]:
```suggestion
Examples:
```SQL
SELECT endsWith('prefix suffix', 'suffix');
SELECT endsWith(blob1, 'suffix') FROM your_dataset;
```
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 8]
```diff
@@ -438,6 +438,86 @@ NOW()
...
+### position
+
+Usage:
+```SQL
+position({needle:string} IN {haystack:string})
+```
+
+Returns the position of one string, `needle`, in another, `haystack`. In SQL, indexes are usually 1-based. That means that position returns 1 if your needle is at the start of the haystack. It only returns 0 if your string is not found.
+
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md:549]:
```suggestion
Examples:
```SQL
SELECT position(':' IN 'hello: world');
SELECT position(':' IN blob1) FROM your_dataset;
```
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 9]
```diff
@@ -438,6 +438,86 @@ NOW()
...
+### substring
+
+Usage:
+```SQL
+substring({string}, {offset:integer}[. {length:integer}]) 
+```
+
+Extracts part of a string, starting at the Unicode code point indicated by the offset and returning the number of code points requested by the length. As previously mentioned, in SQL, indexes are usually 1-based. That means that the offset provided to substring should be at least 1.
+
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md:564]:
```suggestion
Examples:
```SQL
SELECT substring('hello world', 6) AS s;
SELECT substring('hello: world', 1, position(':' IN 'hello: world')-1) AS s;
```
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 10]
```diff
@@ -438,6 +438,86 @@ NOW()
...
+### format
+
+Usage:
+```SQL
+format({string}[, ...])
+```
+
+This function supports formatting strings, integers, floats, datetimes, intervals, etc, except NULL. We do not support literal `{ and }` in the format string and numbered arguments.
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md:578]:
```suggestion
Examples:
```SQL
SELECT format('blob1: {}', blob1) AS s FROM dataset;
```
```

**Comment by 46bit** [comment]
This generally looks really good 👍🏻 I've left some comments with usage examples to make sure people can see how to get started.

File: content/analytics/analytics-engine/sql-reference.md [hunk 11]
```diff
@@ -438,6 +438,138 @@ NOW()
...
+```
+
+### isEmpty
+
+Usage:
+```SQL
+isEmpty({string})
+```
+
+Returns a boolean saying whether the string was empty. This computation can also be done as a binary operation, `{string} = ''`.
```
Comment (pedrosousa) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
Returns a boolean saying whether the string was empty. This computation can also be done as a binary operation: `{string} = ''`.
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 12]
```diff
@@ -438,6 +438,138 @@ NOW()
...
+```
+
+### toUpper
+
+Usage:
+```SQL
+toUpper({string})
+```
+
+Returns the string converted to uppercase. This function is Unicode compatible. This may not be perfect for all languages and users with stringent needs, should do the operation in their own code.
```
Comment (pedrosousa) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
Returns the string converted to uppercase. This function is Unicode compatible. The results may not be perfect for all languages and users with strict needs. These users should do the operation in their own code.
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 13]
```diff
@@ -438,6 +438,138 @@ NOW()
...
+```
+
+Returns a boolean of whether the first string has the second string at its start.
+
+Examples:
+```SQL
+SELECT startsWith('prefix ...', 'prefix');
+SELECT startsWith(blob1, 'prefix') FROM your_dataset;
+```
+### endsWith
```
Comment (pedrosousa) [inline content/analytics/analytics-engine/sql-reference.md:520]:
```suggestion

### endsWith
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 14]
```diff
@@ -438,6 +438,138 @@ NOW()
...
+```
+
+### position
+
+Usage:
+```SQL
+position({needle:string} IN {haystack:string})
+```
+
+Returns the position of one string, `needle`, in another, `haystack`. In SQL, indexes are usually 1-based. That means that position returns 1 if your needle is at the start of the haystack. It only returns 0 if your string is not found.
```
Comment (pedrosousa) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
Returns the position of one string, `needle`, in another, `haystack`. In SQL, indexes are usually 1-based. That means that position returns `1` if your needle is at the start of the haystack. It only returns `0` if your string is not found.
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 15]
```diff
@@ -438,6 +438,138 @@ NOW()
...
+```
+
+### substring
+
+Usage:
+```SQL
+substring({string}, {offset:integer}[. {length:integer}]) 
+```
+
+Extracts part of a string, starting at the Unicode code point indicated by the offset and returning the number of code points requested by the length. As previously mentioned, in SQL, indexes are usually 1-based. That means that the offset provided to substring should be at least 1.
```
Comment (pedrosousa) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
Extracts part of a string, starting at the Unicode code point indicated by the offset and returning the number of code points requested by the length. As previously mentioned, in SQL, indexes are usually 1-based. That means that the offset provided to substring should be at least `1`.
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 16]
```diff
@@ -438,6 +438,138 @@ NOW()
...
+```
+
+### format
+
+Usage:
+```SQL
+format({string}[, ...])
+```
+
+This function supports formatting strings, integers, floats, datetimes, intervals, etc, except NULL. We do not support literal `{` and `}` characters in the format string.
```
Comment (pedrosousa) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
This function supports formatting strings, integers, floats, datetimes, intervals, etc, except `NULL`. The function does not support literal `{` and `}` characters in the format string.
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 17]
```diff
@@ -438,6 +438,138 @@ NOW()
 
 Returns the current time as a DateTime.
 
+### length
```
Comment (pedrosousa) [inline content/analytics/analytics-engine/sql-reference.md:445]: The already documented functions appear in uppercase; the new ones are all in lower/mixed case. Is this difference relevant to the user?
If it's not, I'd rather make them all capitalized (e.g., `LENGTH`). You'd have to update the examples also.

**Review by pedrosousa — APPROVED** [review]

File: content/analytics/analytics-engine/sql-reference.md (hunk 17 above)
Comment (angelampcosta) [reply content/analytics/analytics-engine/sql-reference.md:445]: @46bit could you please confirm if the functions should be capitalized.

Comment (46bit) [reply content/analytics/analytics-engine/sql-reference.md:445]: @angelampcosta It's probably best to capitalise them, yes.

Comment (angelampcosta) [reply content/analytics/analytics-engine/sql-reference.md:445]: Also in the examples, right?

Comment (46bit) [reply content/analytics/analytics-engine/sql-reference.md:445]: Oh of course, there's a lot to change. So, since we're having this conversation...

Function names aren't case-sensitive. I think we probably shouldn't capitalise the function names. It makes them a bit harder to read (e.g. `QUANTILEWEIGHTED` vs `quantileWeighted`.) But undoing that involves a whole bunch of changes, and I haven't had time to go through and make a big PR.

I think it might be best to ship this with functions not uppercased, and then later go through and make the other functions not uppercased. What do people think?

Comment (angelampcosta) [reply content/analytics/analytics-engine/sql-reference.md:445]: I agree. Let's ship these not uppercased and I'll create a ticket for myself to correct the other functions that are now capitalized.

Comment (pedrosousa) [reply content/analytics/analytics-engine/sql-reference.md:445]: I think it makes sense to mention somewhere that function names are not case sensitive (readers may wonder...).

**Comment by ZahoorNoor** [comment]
Classified
Unclassified

File: content/analytics/analytics-engine/sql-reference.md [hunk 18]
```diff
@@ -438,6 +442,139 @@ NOW()
...
+```SQL
+length({string})
+```
+
+Returns the length of a string. This function is UTF-8 compatible.
+
+Examples:
+```SQL
+SELECT length('a string');
+SELECT length(blob1) FROM your_dataset;
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
SELECT length('a string') AS s;
SELECT length(blob1) AS s FROM your_dataset;
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 19]
```diff
@@ -438,6 +442,139 @@ NOW()
...
+```SQL
+isEmpty({string})
+```
+
+Returns a boolean saying whether the string was empty. This computation can also be done as a binary operation: `{string} = ''`.
+
+Examples:
+```SQL
+SELECT isEmpty('a string');
+SELECT isEmpty(blob1) FROM your_dataset;
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
SELECT isEmpty('a string') AS b;
SELECT isEmpty(blob1) AS b FROM your_dataset;
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 20]
```diff
@@ -438,6 +442,139 @@ NOW()
...
+```SQL
+toLower({string})
+```
+
+Returns the string converted to lowercase. This function is Unicode compatible. This may not be perfect for all languages and users with stringent needs, should do the operation in their own code.
+
+Examples:
+```SQL
+SELECT toLower('STRING TO DOWNCASE');
+SELECT toLower(blob1) FROM your_dataset;
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
SELECT toLower('STRING TO DOWNCASE') AS s;
SELECT toLower(blob1) AS s FROM your_dataset;
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 21]
```diff
@@ -438,6 +442,139 @@ NOW()
...
+```SQL
+toUpper({string})
+```
+
+Returns the string converted to uppercase. This function is Unicode compatible. The results may not be perfect for all languages and users with strict needs. These users should do the operation in their own code.
+
+Examples:
+```SQL
+SELECT toUpper('string to uppercase');
+SELECT toUpper(blob1) FROM your_dataset;
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
SELECT toUpper('string to uppercase') AS s;
SELECT toUpper(blob1) AS s FROM your_dataset;
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 22]
```diff
@@ -438,6 +442,139 @@ NOW()
...
+```SQL
+startsWith({string}, {string})
+```
+
+Returns a boolean of whether the first string has the second string at its start.
+
+Examples:
+```SQL
+SELECT startsWith('prefix ...', 'prefix');
+SELECT startsWith(blob1, 'prefix') FROM your_dataset;
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
SELECT startsWith('prefix ...', 'prefix') AS b;
SELECT startsWith(blob1, 'prefix') AS b FROM your_dataset;
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 23]
```diff
@@ -438,6 +442,139 @@ NOW()
...
+```SQL
+endsWith({string}, {string})
+```
+
+Returns a boolean of whether the first string contains the second string at its end.
+
+Examples:
+```SQL
+SELECT endsWith('prefix suffix', 'suffix');
+SELECT endsWith(blob1, 'suffix') FROM your_dataset;
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
SELECT endsWith('prefix suffix', 'suffix') AS b;
SELECT endsWith(blob1, 'suffix') AS b FROM your_dataset;
```

File: content/analytics/analytics-engine/sql-reference.md [hunk 24]
```diff
@@ -438,6 +442,139 @@ NOW()
...
+```SQL
+position({needle:string} IN {haystack:string})
+```
+
+Returns the position of one string, `needle`, in another, `haystack`. In SQL, indexes are usually 1-based. That means that position returns `1` if your needle is at the start of the haystack. It only returns `0` if your string is not found.
+
+Examples:
+```SQL
+SELECT position(':' IN 'hello: world');
+SELECT position(':' IN blob1) FROM your_dataset;
```
Comment (46bit) [inline content/analytics/analytics-engine/sql-reference.md]:
```suggestion
SELECT position(':' IN 'hello: world') AS p;
SELECT position(':' IN blob1) AS p FROM your_dataset;
```

**Review by 46bit — COMMENTED** [review]
I've left a few corrections to my own usage examples, so that they all run when copy-and-pasted. Good to go otherwise :shipit:

**Review by 46bit — APPROVED** [review]

**Comment by angelampcosta** [comment]
Thank you so much @46bit!

**Comment by 46bit** [comment]
Thank you too @angelampcosta -- it's been a real help not having to do this myself
